test/cache_test.lua

```diff
@@ -0,0 +1,163 @@
+        temporary = false,
+        format = table.deepcopy(helper.test_space_format())
+    }
+    table.insert(g.space.format, 1, {
+        name = 'bucket_id', type = 'unsigned', is_nullable = false
+    })
+
+    g.space.indexes = {
+        table.deepcopy(primary_index),
+        table.deepcopy(bucket_id_idx)
+    }
+    g.space.sharding_key = {'unsigned_nonnull', 'integer_nonnull'}
+    g.schema = {
+        spaces = {
+            space = g.space,
+        }
+    }
+end
+
+local g = t.group()
+g.before_all(db.init)
+g.before_each(space_init)
+
+function g.test_cache_processed_func_body()
+    g.schema.spaces.space.sharding_func = {
+        body = func_body_first
+    }
+    local ok, err = ddl.set_schema(g.schema)
+    t.assert_equals(err, nil)
+    t.assert_equals(ok, true)
+
+    local res = cache.internal.get('space')
+    t.assert(res)
+    t.assert(res.callable)
+    t.assert_type(res.callable, 'function')
+    t.assert_equals(res.callable(), 42)
+end
+
+function g.test_cache_processed_func_name()
+    local sharding_func_name = 'sharding_func'
+    rawset(_G, sharding_func_name, function(key) return key end)
+    g.schema.spaces.space.sharding_func = sharding_func_name
+
+    local ok, err = ddl.set_schema(g.schema)
+    t.assert_equals(err, nil)
+    t.assert_equals(ok, true)
+
+    local res = cache.internal.get('space')
+    t.assert(res)
+    t.assert(res.parsed_func_name)
+    t.assert_type(res.parsed_func_name, 'table')
+    t.assert_equals(res.parsed_func_name[1], 'sharding_func')
+
+    rawset(_G, sharding_func_name, nil)
+end
+
+function g.test_cache_func_body_changed()
+    local sharding_func_name = 'sharding_func'
+    rawset(_G, sharding_func_name, function(key) return key + 10 end)
+    g.schema.spaces.space.sharding_func = sharding_func_name
+
+    local ok, err = ddl.set_schema(g.schema)
+    t.assert_equals(err, nil)
+    t.assert_equals(ok, true)
+
+    local id, err = ddl.bucket_id('space', 42)
+    t.assert_equals(err, nil)
+    t.assert_type(id, 'number')
+    t.assert_equals(id, 52)
+
+    rawset(_G, sharding_func_name, function(key) return key + 11 end)
+    local id, err = ddl.bucket_id('space', 42)
+    t.assert_equals(err, nil)
+    t.assert_type(id, 'number')
+    t.assert_equals(id, 53)
+
+    rawset(_G, sharding_func_name, nil)
+end
+
+function g.test_cache_schema_changed()
+    g.schema.spaces.space.sharding_func = {
+        body = func_body_first
+    }
+    local ok, err = ddl.set_schema(g.schema)
+    t.assert_equals(err, nil)
+    t.assert_equals(ok, true)
+
+    local res = cache.internal.get('space')
+    t.assert(res)
+    t.assert(res.raw_tuple)
+    t.assert_equals(res.raw_tuple[SPACE_NAME_IDX], 'space')
+    t.assert_equals(res.raw_tuple[SHARD_FUNC_NAME_IDX], nil)
+    t.assert_equals(res.raw_tuple[SHARD_FUNC_BODY_IDX], func_body_first)
+
+    space_init(g)
+
+    local res = cache.internal.get('space')
+    t.assert_equals(res, nil)
+end
+
+function g.test_cache_space_updated()
```

Comment:
Shouldn't we check what the callable will return as well?

Reply:
agree

Reply:
This test case just check that body was changed after space updated (trigger works).

Reply:
fixed